types/src/transaction/mod.rs

```diff
@@ -738,6 +739,47 @@ impl TransactionWithProof {
     }
 }
 
+/// The status of VM execution, which contains more detailed failure info
+#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash, Serialize, Deserialize)]
+#[cfg_attr(any(test, feature = "fuzzing"), derive(Arbitrary))]
+#[cfg_attr(any(test, feature = "fuzzing"), proptest(no_params))]
+pub enum KeptTXNStatus {
```

Comment:
can we just call this TransactionExecutionStatus?

Reply:
after checking the error info https://github.com/areshand/move/blob/0747a72af49e7116ae2b54625668d8536ee086d8/language/move-core/types/src/vm_status.rs#L491-L598, and here https://github.com/areshand/move/blob/0747a72af49e7116ae2b54625668d8536ee086d8/language/move-core/types/src/vm_status.rs#L624-L642. These errors can be detected when user submitting the txn. There is no need to keep these txn at the first place. As a followup of this issue, we should directly invalidate the request when user submit the txn. will have a follow up PR to (1) invalidate the txn submit request (2) copy the same VM validation logic in early stage and return the detailed error info.

Reply:
If it's already persisted status, I think we can skip the Kept prefix

Reply:
modified the name to "TransExecutionStatus"

Reply:
typically abbr for transaction is txn not trans, also I'd like shorter name, maybe "ExecutionStatus"? naming is hard

Reply:
> These errors can be detected when user submitting the txn. There is no need to keep these txn at the first place. 
I think the status code we are gonna include as part of the transaction is better to be of only "kept" status. A code that indicates something that's rejected from the mempool should not be part of this, I assume? Did I miss anything?

Reply:
The name of this one doesn't have "Kept" in it is fine, but if through the API someone asks the status of a transaction, we should still return some enum from which the user can tell if the txn is discarded by validation, or discarded by vm, or accepted by the chain but aborted, or succeeded.

Reply:
tl;dr: user won't confuse about the failure status

A failed txn would fall into the following groups: 
1. mempool validation check signature and prologue. The error message explicitly shows this is invalid txn https://github.com/areshand/aptos-core-1/blob/18edbd1687378389ff172384fd382db9a4ede4cf/api/src/transactions.rs#L205
2. VM execution
	a. kept with error: this is the ExecutionStatus we introduced to provide more detailed error. ExecutionStatus is not used in 1. or 2.b scenarios. But the error msg definitely need more improvement to let user know their txn are committed with error. 
	b. discard: this is interesting. the user would still see the txn executed. But they won't be able to query the txn since these txns are not committed. Ideally, these txns should be completely handled in mempool validation. But this require more investigation.